### ISSUE TITLE
Add comment-*-marker-newline-* rules

### ISSUE THREAD

**Issue by alexander-akait** [stylelint/stylelint#1516]
- [ ] `comment-opening-marker-newline-before`
- [ ] `comment-closing-marker-newline-after`

Valid

``` css
 /* Comment */
div,
 /* Comment */,
.class {
  /* Comment */
  display: inline;
  /* Comment */
}
 /* Comment */
```

and etc.

Invalid:

``` css
div {
  display: inline; /* Some comment */
}
```

and etc.

**Comment by jeddy3** [comment]
Sgtm.

- Name: `comment-no-inline`
- Primary option: `true`
- Secondary options: none
- Message: "Unexpected inline comment"
- Section: "Comment"

Does anyone have any queries about this, or shall I label it was as ready?

**Comment by ntwb** [comment]
Need to ensure compatibility with the `max-line-length` rule and `ignore: ["non-comments"]` option
https://github.com/stylelint/stylelint/tree/master/src/rules/max-line-length#ignore-non-comments

**Comment by jeddy3** [comment]
I think that any conflict that arises from turning on the rule in this proposal and `max-line-length` rule with its `ignore: ["non-comments"]` option [will need to be resolved by the user](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/faq.md#how-do-i-manage-conflicts-between-rules).

Any other queries, or shall I label up?

**Comment by davidtheclark** [comment]
> will need to be resolved by the user.

Agreed.

**Comment by ntwb** [comment]
All good 👍

**Comment by davidtheclark** [comment]
Thought about it another minute and I'm not thriled about the name `comment-no-inline`, because "inline comment" is used to refer to Sass/Less-style `//` comments (e.g. https://en.wikipedia.org/wiki/Comment_(computer_programming)). Can we think of another name?

**Comment by ntwb** [comment]
Pretty much there are two comment types, "block" or "inline" (also referred to as "line")

Broadly speaking both the comments below in programming languages both the comments in this code are referred to as `inline` comments...
```css

div {
  /* Some comment */
  display: inline; /* Some comment */
}
```

I searched through that Wikipedia reference and the one below, hunting for a generised term relating to "shared", "same line", "end of line" etc and nothing really stood out, the only thing I came up with was `comment-no-end-of-line` though I don't think that is the exactly succinct either.
https://en.wikipedia.org/wiki/Comparison_of_programming_languages_(syntax)#Comments

**Comment by m-allanson** [comment]
Would these be considered warnings?

```css
div {
  /* comment */ display: block;
}

div {
  display: /* comment */ block;
}
```

If so, the rule is to prevent a comment from appearing on the same line as any CSS statement. Maybe `comment-always-newlines` or `comment-no-shared-line` would work as the rule name?

**Comment by jeddy3** [comment]
`comment-no-shared-line` sounds like the best so far to me.

**Comment by jeddy3** [comment]
I've come to realise that when we struggle to name something (`*-no-empty-line-before`, `blockless-groups` etc) it's generally because we veering off our usual path.

I think what we're trying to achieve with this rule can be done with our usual `*-newline-before/after` approach. It'll be more powerful too.

- `comment-opening-marker-newline-before`
- `comment-closing-marker-newline-after`

Want to disallow shared line comments? e.g.

```css
div {
  /* comment */ display: block;
  display: /* comment */ block;
  display: block;  /* comment */
}
```

Use:

```js
"comment-opening-marker-newline-before": "always"
"comment-closing-marker-newline-after": "always"
```

Want to allow end of line comments but disallow comments like this?:

```css
div {
  /* comment */ display: block;
  display: /* comment */ block;
}
```

```js
"comment-closing-marker-newline-after": "always"
```

We can take it further and add:

- `comment-opening-marker-space-before`

Want to allow comments on lines by themselves and comments at the end of lines (but with single space before it) e.g.

```css
div {
  /* comment */
  display: block; /* comment */
}
```

```js
"comment-opening-marker-space-before": "always-shared-line"
"comment-closing-marker-newline-after": "always"
```

Sound good?

**Comment by davidtheclark** [comment]
Keep in mind: as soon as this rule is released people are going to request exceptions for `//`-comments.

@jeddy3 At first at first my gut reacted against this new proposal ... but I think my gut may have been wrong. I personally would not use rules restricting comment usage, since part of their glory is their freedom to be inserted anywhere; but if people want to start writing rules to restrict comma usage, I think what you've proposed is the way to go.

**Comment by jeddy3** [comment]
> Keep in mind: as soon as this rule is released people are going to request exceptions for //-comments.

In the existing comment whitespace rule, we [ignore `//` comments outright](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-whitespace-inside/index.js#L30). And push supporting `//` comments to plugins (https://github.com/kristerkari/stylelint-scss/issues/48). I figured we'd do the same for this.

> since part of their glory is their freedom

I agree with this sentiment and I won't be writing these rules myself :)

**Comment by ntwb** [comment]
I was thinking the rule names even require `marker` then realised we already have the [comment-empty-line-before](https://github.com/stylelint/stylelint/tree/master/src/rules/comment-empty-line-before) rule, wouldn't an equivalent rule `comment-empty-line-after` be more appropriate than `"comment-closing-marker-newline-after": "always"` ?

**Comment by jeddy3** [comment]
> wouldn't an equivalent rule `comment-empty-line-after` be more appropriate than `"comment-closing-marker-newline-after": "always"` ?

An `empty-line` and a `newline` are two very different things.

Take these three examples:

```js
{
  "comment-empty-line-before": "always",
  "comment-open-marker-newline-before": "always"
}
```

```css
@import "x.css";

/* comment on a newline with an empty line before */
```

Vs

```js
{
  "comment-empty-line-before": "never",
  "comment-open-marker-newline-before": "always"
}
```

```css
@import "x.css";
/* comment on a newline without an empty line before */
```

Vs

```js
{
  "comment-open-marker-newline-before": "never"
}
```

```css
@import "x.css"; /* comment on a shared-line */
```

As an aside, we only have `empty-line-before` rules, and no `empty-line-after` rules. This is to avoid conflict and for consistency. Just like in CSS when a team decides to only using `margin-top`, rather than `margin-bottom` as well. A developer only needs to remember that empty lines are controlled by *pushing* the *current thing* away from the *preceding thing*, rather than the *succeeding thing*:

```js
{
  "at-rule-empty-line-before": "always",
  "comment-empty-line-before": "always",
  "comment-open-marker-newline-before": "always"
}
```

```css
@import "x.css";

/* 
 * comment on a newline with an empty line before 
 * the succeeding at-custom-media is responsible for pushing itself away from the this comment.
 * Just in the same way that this comment is responsible for pushing itself away from the
 * the preceding at-import.
*/

@custom-media --small-viewport (max-width: 30em);
```

Does that make more sense?

I hope to find time to expand the About rules doc's [whitespace](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md#whitespace-rules) and [rules work together](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md#rules-work-together) sections to include something more about these `empty-line-before` rules as their working is quite subtle :)

**Comment by jeddy3** [comment]
Any more reservations, or shall I convert this issue to?:

- `comment-opening-marker-newline-before`
- `comment-closing-marker-newline-after`

**Comment by evilebottnawi** [comment]
@jeddy3 sgtm

**Comment by jeddy3** [comment]
- Names:
  - `comment-opening-marker-newline-before`
  - `comment-closing-marker-newline-after`
- Primary options: `"always"|"always-shared-line"`
- Secondary options: none
- Messages: 
  - `Expected newline before "/*"`.
  - `Expected newline after "*/"`.
- Section: "Comment"

**Comment by evilebottnawi** [comment]
@jeddy3 can give example with `always-shared-line` option?

**Comment by davidtheclark** [comment]
@evilebottnawi @jeddy3 Is this still under consideration?

**Comment by evilebottnawi** [comment]
@davidtheclark for me yes

**Comment by jeddy3** [comment]
> @evilebottnawi @jeddy3 Is this still under consideration?

I don't think there's anything _technically_ wrong with the proposal. However, I'm not interested in building it though because, as you said a while back, the beauty of comments are that you can throw them in anywhere (read: the most appropriate place).

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to [create plugins](https://stylelint.io/developer-guide/plugins).
